### PR TITLE
[Add] Support swap the order of convolution module and self-attention module.

### PR DESCRIPTION
**Pull request by TeaPoly**
Reference: A Better and Faster End-to-End Model for Streaming ASR, https://arxiv.org/abs/2011.10798

It's very close to Conformer result. 

Only setting yaml config `pos_enc_layer_type: 'conv_module_pos'`.

## FastConformer Result
 
  * Feature info: using fbank feature, dither, cmvn, online speed perturb
  * Training info: lr 0.002, batch size 16, 4 gpu, acc_grad 4, 240 epochs, dither 0.1
  * Decoding info: ctc_weight 0.5, average_num 20
 
  | decoding mode             | CER   |
  |---------------------------|-------|
  | attention decoder         | 5.67  |
  | ctc greedy search         | 4.96  |
  | ctc prefix beam search    | 4.96  |
  | attention rescoring       | 4.66  |

## Conformer Result

* Feature info: using fbank feature, dither, cmvn, online speed perturb
* Training info: lr 0.002, batch size 18, 4 gpu, acc_grad 4, 240 epochs, dither 0.1
* Decoding info: ctc_weight 0.5, average_num 20

| decoding mode             | CER   |
|---------------------------|-------|
| attention decoder         | 5.18  |
| ctc greedy search         | 4.94  |
| ctc prefix beam search    | 4.94  |
| attention rescoring       | 4.61  |

## Transformer Result

* Feature info: using fbank feature, dither, with cmvn, online speed perturb.
* Training info: lr 0.002, batch size 26, 4 gpu, acc_grad 4, 240 epochs, dither 0.1
* Decoding info: ctc_weight 0.5, average_num 20

| decoding mode             | CER   |
|---------------------------|-------|
| attention decoder         | 5.69  |
| ctc greedy search         | 5.92  |
| ctc prefix beam search    | 5.91  |
| attention rescoring       | 5.30  |

### CODE REVIEW

**Comment by TeaPoly** [comment]
The result is as follows,  and Confomer and Transformer result is copy from examples/aishell/s0）：

| **decoding mode**      | **Conformer** | **FastConformer** | **Transformer** |
| ---------------------- | ------------- | ----------------- | --------------- |
| attention decoder      | 5.18          | 5.67              | 5.69            |
| ctc greedy search      | 4.94          | 4.96              | 5.92            |
| ctc prefix beam search | 4.94          | 4.96              | 5.91            |
| attention rescoring    | 4.61          | 4.66              | 5.30            |

File: wenet/transformer/encoder.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 from wenet.transformer.convolution import ConvolutionModule
 from wenet.transformer.embedding import PositionalEncoding
 from wenet.transformer.embedding import RelPositionalEncoding
+from wenet.transformer.embedding import NoPositionalEncoding
```
Comment (whiteshirt0429) [inline wenet/transformer/encoder.py:17]: use  from wenet.transformer.embedding import (PositionalEncoding, RelPositionalEncoding,NoPositionalEncoding )

**Review by whiteshirt0429 — COMMENTED** [review]
very exciting results

**Comment by whiteshirt0429** [comment]
looking forward to u2 results of the faster conformer version

**Comment by TeaPoly** [comment]
> looking forward to u2 results of the faster conformer version

## U2 FastConformer
- Feature info: using fbank feature, dither, cmvn, online speed perturb
- Training info: lr 0.002, batch size 16, 4 gpu, acc_grad 4, 240 epochs, dither 0.1
- Decoding info: ctc_weight 0.5, average_num 20

decoding mode/chunk size | full | 16 | 8 | 4
-- | -- | -- | -- | --
attention decoder | 5.31 | 5.56 | 5.60 | 5.72
ctc greedy search | 5.54 | 6.22 | 6.56 | 6.95
ctc prefix beam search | 5.55 | 6.22 | 6.56 | 6.95
attention rescoring | 4.95 | 5.41 | 5.61 | 5.83

## U2 Conformer
- Feature info: using fbank feature, dither=0, cmvn, oneline speed perturb
- Training info: lr 0.001, batch size 16, 8 gpu, acc_grad 1, 180 epochs, dither 0.0
- Decoding info: ctc_weight 0.5, average_num 20

decoding mode/chunk size | full | 16 | 8 | 4
-- | -- | -- | -- | --
attention decoder | 5.40 | 5.60 | 5.74 | 5.86
ctc greedy search | 5.56 | 6.29 | 6.68 | 7.10
ctc prefix beam search | 5.57 | 6.30 | 6.67 | 7.10
attention rescoring | 5.05 | 5.45 | 5.69 | 5.91

**Comment by fanlu** [comment]
@TeaPoly Could you please upload the U2 FastConformer yaml.

> > looking forward to u2 results of the faster conformer version
> 
> ## U2 FastConformer
> * Feature info: using fbank feature, dither, cmvn, online speed perturb
> * Training info: lr 0.002, batch size 16, 4 gpu, acc_grad 4, 240 epochs, dither 0.1
> * Decoding info: ctc_weight 0.5, average_num 20
> 
> decoding mode/chunk size	full	16	8	4
> attention decoder	5.31	5.56	5.60	5.72
> ctc greedy search	5.54	6.22	6.56	6.95
> ctc prefix beam search	5.55	6.22	6.56	6.95
> attention rescoring	4.95	5.41	5.61	5.83
> ## U2 Conformer
> * Feature info: using fbank feature, dither=0, cmvn, oneline speed perturb
> * Training info: lr 0.001, batch size 16, 8 gpu, acc_grad 1, 180 epochs, dither 0.0
> * Decoding info: ctc_weight 0.5, average_num 20
> 
> decoding mode/chunk size	full	16	8	4
> attention decoder	5.40	5.60	5.74	5.86
> ctc greedy search	5.56	6.29	6.68	7.10
> ctc prefix beam search	5.57	6.30	6.67	7.10
> attention rescoring	5.05	5.45	5.69	5.91

**Comment by TeaPoly** [comment]
> @TeaPoly Could you please upload the U2 FastConformer yaml.
> 
> > > looking forward to u2 results of the faster conformer version
> > 
> > 
> > ## U2 FastConformer
> > 
> > * Feature info: using fbank feature, dither, cmvn, online speed perturb
> > * Training info: lr 0.002, batch size 16, 4 gpu, acc_grad 4, 240 epochs, dither 0.1
> > * Decoding info: ctc_weight 0.5, average_num 20
> > 
> > decoding mode/chunk size	full	16	8	4
> > attention decoder	5.31	5.56	5.60	5.72
> > ctc greedy search	5.54	6.22	6.56	6.95
> > ctc prefix beam search	5.55	6.22	6.56	6.95
> > attention rescoring	4.95	5.41	5.61	5.83
> > ## U2 Conformer
> > 
> > * Feature info: using fbank feature, dither=0, cmvn, oneline speed perturb
> > * Training info: lr 0.001, batch size 16, 8 gpu, acc_grad 1, 180 epochs, dither 0.0
> > * Decoding info: ctc_weight 0.5, average_num 20
> > 
> > decoding mode/chunk size	full	16	8	4
> > attention decoder	5.40	5.60	5.74	5.86
> > ctc greedy search	5.56	6.29	6.68	7.10
> > ctc prefix beam search	5.57	6.30	6.67	7.10
> > attention rescoring	5.05	5.45	5.69	5.91

It's not merged. You can found in "Files Changed".

File: examples/aishell/s0/conf/train_unified_fastconformer.yaml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    sort: true
+
+grad_clip: 5
+accum_grad: 1
+max_epoch: 240
+log_interval: 100
+
+optim: adam
+optim_conf:
+    lr: 0.001
```
Comment (fanlu) [inline examples/aishell/s0/conf/train_unified_fastconformer.yaml:87]: this yaml is not the same as the trainint info you posted before?
`Training info: lr 0.002, batch size 16, 4 gpu, acc_grad 4, 240 epochs, dither 0.1`

Comment (TeaPoly) [reply examples/aishell/s0/conf/train_unified_fastconformer.yaml:87]: I have modify the result. yaml is correct.  I only change `max_epoch`, `pos_enc_layer_type` and `selfattention_layer_type` in file `examples/aishell/s0/train_unified_conformer.yaml`.

**Review by fanlu — APPROVED** [review]

**Comment by robin1001** [comment]
@TeaPoly , I added NoPositionalEncoding recently. It is pretty simple in my implementation and there is no convolution swap, here is the PR https://github.com/wenet-e2e/wenet/pull/405 and result, we can see the it got promising result without PE.

| decoding mode          | with PE | without PE |
|------------------------|---------|------------|
| attention decoder      | 5.18    | 5.73       |
| ctc greedy search      | 4.94    | 4.97       |
| ctc prefix beam search | 4.94    | 4.97       |
| attention rescoring    | 4.61    | 4.69       |

So I wonder what does "fast" mean? and maybe there is no need to swap the convolution and attention in Conformer.

**Comment by TeaPoly** [comment]
> @TeaPoly , I added NoPositionalEncoding recently. It is pretty simple in my implementation and there is no convolution swap, here is the PR #405 and result, we can see the it got promising result without PE.
> 
> decoding mode	with PE	without PE
> attention decoder	5.18	5.73
> ctc greedy search	4.94	4.97
> ctc prefix beam search	4.94	4.97
> attention rescoring	4.61	4.69
> So I wonder what does "fast" mean? and maybe there is no need to swap the convolution and attention in Conformer.

I use 'fast' conformer is inspired from title of paper "A Better and Faster End-to-End Model for Streaming ASR", it means no need relative position encoding. 
And they using features following: 
> Features from 4 contiguous frames are stacked to form a 512D input representation, which is further sub- sampled by a factor of 3.

Not using conv-subsamping. I guess swap operation it work with inputs features without relative position information.

**Comment by TeaPoly** [comment]
The results of the experiment are reserved for reference.

**Pull request closed without merging**